websocket/connection.go

```diff
@@ -0,0 +1,176 @@
+	defaultPingWait time.Duration
+	// Channel to stop the go routines
+	done chan bool
+}
+
+//Websocket connection routine. setup the ping-pong and connection settings
+func NewConnection(serverURL string, token string) (*Connection, error) {
+	wsServerURL := "ws" + strings.TrimRight(serverURL[4:], "/")
+	parsedURL, err := url.Parse(wsServerURL + defaultDeviceConnectPath)
+	if err != nil {
+		return nil, err
+	}
+
+	wslib.DefaultDialer.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: true,
+	}
+	var wsconn *wslib.Conn
+	dialer := *wslib.DefaultDialer
+
+	headers := http.Header{}
+	headers.Set("Authorization", "Bearer "+token)
+	wsconn, resp, err := dialer.Dial(parsedURL.String(), headers)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	c := &Connection{
+		connection:      wsconn,
+		writeWait:       writeWait,
+		maxMessageSize:  maxMessageSize,
+		defaultPingWait: defaultPingWait,
+		done:            make(chan bool),
+	}
+	wsconn.SetReadLimit(maxMessageSize)
+	go c.pingPongHandler()
+
+	return c, nil
+}
+
+func (c *Connection) pingPongHandler() {
+	// handle the ping-pong connection health check
+	err := c.connection.SetReadDeadline(time.Now().Add(c.defaultPingWait))
+	if err != nil {
+		return
+	}
+
+	pingPeriod := (c.defaultPingWait * 9) / 10
+	ticker := time.NewTicker(pingPeriod)
+	defer ticker.Stop()
+
+	c.connection.SetPongHandler(func(string) error {
+		ticker.Reset(pingPeriod)
+		return c.connection.SetReadDeadline(time.Now().Add(c.defaultPingWait))
+	})
+
+	c.connection.SetPingHandler(func(msg string) error {
+		ticker.Reset(pingPeriod)
+		err := c.connection.SetReadDeadline(time.Now().Add(c.defaultPingWait))
+		if err != nil {
+			return err
+		}
+		c.writeMutex.Lock()
+		defer c.writeMutex.Unlock()
+		return c.connection.WriteControl(
+			wslib.PongMessage,
+			[]byte(msg),
+			time.Now().Add(c.writeWait),
+		)
+	})
+
+	running := true
+	for running {
+		select {
+		case <-c.done:
+			running = false
+			break
+		case <-ticker.C:
+			pongWaitString := strconv.Itoa(int(c.defaultPingWait.Seconds()))
+			c.writeMutex.Lock()
+			_ = c.connection.WriteControl(
+				wslib.PingMessage,
+				[]byte(pongWaitString),
+				time.Now().Add(c.defaultPingWait),
+			)
+			c.writeMutex.Unlock()
+		}
+	}
+}
+
+func (c *Connection) GetWriteTimeout() time.Duration {
+	return c.writeWait
+}
+
+func (c *Connection) WriteMessage(m *ws.ProtoMsg) (err error) {
+	data, err := msgpack.Marshal(m)
+	if err != nil {
+		return err
+	}
+	c.writeMutex.Lock()
+	defer c.writeMutex.Unlock()
+	_ = c.connection.SetWriteDeadline(time.Now().Add(c.writeWait))
+	return c.connection.WriteMessage(wslib.BinaryMessage, data)
+}
+
+func (c *Connection) ReadMessage() (*ws.ProtoMsg, error) {
+	_, data, err := c.connection.ReadMessage()
+	if err != nil {
+		return nil, err
+	}
+
+	m := &ws.ProtoMsg{}
+	err = msgpack.Unmarshal(data, m)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (c *Connection) Close() error {
```

Comment:
Nitpick: The client does not send a close control frame if it's terminating the connection. However, since we don't care from deviceconnect point of view it's also not necessary.